File: java/ankit_kanyal/devnest_problems/RemoveDuplicateLetters.java
// link- https://devsnest.in/algo-challenges/remove-duplicate-letters

// discussion- https://devsnest.in/algo-challenges/remove-duplicate-letters

static String solve(String s){
//CODE HERE 
    int alphabets[]=new int[26];
    int visited[]=new int[26];

    for (int i = 0; i < 26; i++){
            alphabets[i] = 0;
            visited[i] = 0;      //false
        }
    for(char ch:s.toCharArray()){
        int no=(int)ch-97;
        alphabets[no]+=1;
    }

    String ans="";
    

    for(char ch:s.toCharArray()){
        int index=ch-97;
        alphabets[index]--;


        if(visited[index]==1) continue;

        while(!ans.equals("") && ch<ans.charAt(ans.length()-1) && alphabets[ans.charAt(ans.length()-1)-97]>0){
            visited[ans.charAt(ans.length()-1)-97]=0;
            ans=ans.substring(0,ans.length()-1);
        }
        ans+=ch;
        visited[index]=1;
    }

    return ans;
}

File: py.thon/vedant_bhatnagar/DevsnestProblems_THA/Phase 1/removeDuplicateLetters.py
def solve(s):
    # CODE HERE
    lastindex = {}
    for i, el in enumerate(s):
        lastindex[el] = i

    st = []
    present = set()
    for i, el in enumerate(s):
        if el in present:
            continue
        while st and st[-1] > el and lastindex[st[-1]] > i:
            t = st.pop()
            present.remove(t)
        st.append(el)
        present.add(el)
    return "".join(st)

